### ISSUE TITLE
Setup script defining wrong and missing variables/ using deprecated attributes

### ISSUE THREAD

**Issue by gms1234** [Icinga/icingaweb2-module-director#1504]
### Master:

```
Application information:
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: Ubuntu
  Platform version: 16.04.4 LTS (Xenial Xerus)
  Kernel: Linux
  Kernel version: 4.4.0-042stab127.2
  Architecture: x86_64

Build information:
  Compiler: GNU 5.3.1
  Build host: 9c880c2f42f5

Icinga2 version: r2.8.4-1
Director version: master
Icingaweb2 version: 2.5.3
```

### Client/Agent:

```
Application information:
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: Ubuntu
  Platform version: 16.04.4 LTS (Xenial Xerus)
  Kernel: Linux
  Kernel version: 4.4.0-042stab127.2
  Architecture: x86_64

Build information:
  Compiler: GNU 5.3.1
  Build host: 9c880c2f42f5
  
Icinga2 version: r2.8.4-1
```

### Problem:

I tried using the setup script within the Icinga2 director to configure an agent, but with no success. The script uses wrong variables and doesn't define a zone for the master.

The script uses attributes which (Icinga2 says so itself) are deprecated:

```
warning/ApiListener: Attribute 'key_path' for object 'api' of type 'ApiListener' is deprecated and should not be used.
warning/ApiListener: Attribute 'ca_path' for object 'api' of type 'ApiListener' is deprecated and should not be used.
warning/ApiListener: Attribute 'cert_path' for object 'api' of type 'ApiListener' is deprecated and should not be used.
warning/ApiListener: Please read the upgrading documentation for v2.8: https://www.icinga.com/docs/icinga2/latest/doc/16-upgrading-icinga-2/
```

Also, the script defines two zones for 'icinga-client' which leads to an error because `'icinga-client'` is not allowed to be in two zones.

### Example:

```
ICINGA2_NODENAME='icinga-client'
ICINGA2_CA_TICKET='69dd6ede95235f8768c4ce2ff55cde4027ddbb51'
ICINGA2_PARENT_ZONE='client-zone'
ICINGA2_PARENT_ENDPOINTS='icinga-client'
ICINGA2_CA_NODE='icinga-master'

ZONES_ICINGA2=`cat << EOF
/** Icinga 2 Config - proposed by Icinga Director */

// TODO: improve establish connection handling
object Endpoint "${ICINGA2_NODENAME}" {}
object Endpoint "${ICINGA2_CA_NODE}" {}
object Zone "${ICINGA2_PARENT_ZONE}" {
  endpoints = [ "$ICINGA2_PARENT_ENDPOINTS" ]
  // TODO: all endpoints in master zone
}

object Zone "director-global" { global = true }

object Zone "${ICINGA2_NODENAME}" {
  endpoints = [ "$ICINGA2_NODENAME" ]
  parent = "${ICINGA2_PARENT_ZONE}"
}
EOF
`
```

In my opinion (and after checking in with a colleague who has access to an older version of Icinga2 and director in which the variables are set correctly) `ICINGA2_PARENT_ENDPOINTS` should not say `'icinga-client'` and instead list the `'icinga-master'`. 

It should also define the `'master-zone'` which isn't even mentioned in the script, except for a `// TODO` comment. Is the script even maintained?

### What I tried:

I checked if there was any problem with syncing the names to the corresponding variables by manually editing the zone (within the director) in which the client resides. 

The script changed - so somehow the function which defines the setup script sets `ICINGA2_PARENT_ENDPOINTS` wrong. 

I got it to run by manually changing the variables but this defies the purpose of the setup script.

**Comment by Thomas-Gelf** [comment]
Partially duplicates #1425, problem is that Director still supports Icinga v2.6.x

**Comment by gms1234** [comment]
Will there be a fix in the near future? 

Under "Requirements" it says "It is recommended to use the latest feature release of Icinga 2." But right now it seems to me that this is not actually encouragable. Actually this feature is not at all usable without manually changing the variables.

**Comment by Thomas-Gelf** [comment]
You can ask `git blame`, I'm innocent ;-) Jokes apart, @gms1234: the endpoints/zones part of the issue should be fixed immediately, the code is definitively incomplete. There is #881 - and somehow got forgotten (blame me). Could you please give that one a try?

For the SSL dir -> currently I do not see a problem here apart from the warning at daemon startup time. I prefer not breaking supported versions, even if it is of course "recommended" to use the latest and greatest. Apart from the warning, am I missing a problem with those paths? We could easily fix this by checking the version provided by `icinga2 --version`.

In addition to the above, there could arise issues from Icinga 2 [issue #6157](https://github.com/Icinga/icinga2/issues/6157). @lazyfrosch: given that potential change and related ones like [issue #6328](https://github.com/Icinga/icinga2/issues/6328) - it would be great if you could give some guidance on how to write such a OS-independent script supporting v2.6-v2.9.

Thanks,
Thomas

**Comment by Thomas-Gelf** [comment]
This is a blocker for v1.5.0.

**Comment by gms1234** [comment]
I edited the Icinga2Agent.bash with the propsed code and restarted the Icinga2 service - nothing changed in the setup script. Am I maybe doing something wrong?

I understand your concerns about the paths. I just mentioned them as I was filing a bug report, they are not my main problem. I do not see any problems right now, just that they could potentially break things in the future. 

Thanks for the taking the time!

**Comment by gms1234** [comment]
I checked the script again. It might be I do not fully understand it, but the variable definitions are still weird to me. 

Example: 
```
for PARENTZONE_ENDPOINT in ${ICINGA2_PARENT_ENDPOINTS[@]}
do
ZONES_ICINGA2=`cat <<EOF
$ZONES_ICINGA2
object Endpoint "${PARENTZONE_ENDPOINT}" {}
EOF
`
done
```

First, the Array `ICINGA2_PARENT_ENDPOINTS` is not actually defined anywhere in the code. 

Second, `ZONES_ICINGA2`references itself and is also not defined anywhere. 

Like I said, take this with a grain of salt, I might just not understand the code.

**Comment by Thomas-Gelf** [comment]
The script alone isn't able to run. When you go to a host's `Agent` tab, it shows a complete one based on:
* the header template
* ENV variables rendered by the Director. They are filled as follows:
```
'ICINGA2_NODENAME'         => $this->getCertName(),
'ICINGA2_CA_TICKET'        => $this->getTicket(),
'ICINGA2_PARENT_ZONE'      => $this->getParentZone()->getObjectName(),
'ICINGA2_PARENT_ENDPOINTS' => array_keys($this->getParentEndpoints()),
'ICINGA2_CA_NODE'          => $this->getCaServer(),
```
* the rest of the script

**Comment by Thomas-Gelf** [comment]
So, apart from the issue with multiple parent endpoints (which the linked pull request tries to fix). The script alone without those parameters will not be able to run.

**Comment by gms1234** [comment]
Maybe we are not understanding each other here. I checked the `Agent` tab. I expected the values like `ICINGA2_PARENT_ENDPOINT` to change there. They did not. 

Also, if I understand you correctly, the problem lies within the definition of the variables by the Director? What I mean with that: The variable `ICINGA_PARENT_ZONE` is defined within the Director and given out to the `Icinga2Agent.bash` script. But somehow the way it defines those variables is screwed. 

Am I correct with this assumption?

**Comment by Thomas-Gelf** [comment]
That's at least what happens in my setup, I quickly verified it in a small test lab. Looks like this with the current master and no additional patches:

```bash
#!/bin/bash

# This generates and signs your required certificates. Please do not
# forget to install the Icinga 2 package and your desired monitoring
# plugins first:

ICINGA2_NODENAME='app1.cmdb.example.com'
ICINGA2_CA_TICKET='7ab6815ea5baac6b139384fe578ceabb205d7ba4'
ICINGA2_PARENT_ZONE='master'
ICINGA2_PARENT_ENDPOINTS='i2-master1.example.com'
ICINGA2_CA_NODE='i2-master1.example.com'
# Make sure icinga2 is installed and running

[..]
```

**Comment by gms1234** [comment]
Is this with or without the changed script? 

Also, my master-zone is named differently. Might it be a problem that it is not the default name?

**Comment by Thomas-Gelf** [comment]
Without the changed script. Changing the master zone should not be a problem at all. Director fills above variables - based on it's given knowledge about your Zones and Endpoints.

**Comment by gms1234** [comment]
This is odd. I checked it on a newly set up system myself and the problem occurs there, too.

**Comment by Thomas-Gelf** [comment]
Could you please show me a sample host, it's config preview and generated Agent Script header? Eventually anonymize host name and other sensible information, but please in a way allowing me to still understand involved relations ;-)

**Comment by gms1234** [comment]
object Host "icinga-client" {
    display_name = "icinga-client"
    address = "icinga-client-fqdn"
    check_command = "hostalive"
    max_check_attempts = "3"
    check_interval = 3m
    retry_interval = 1m
    check_timeout = 1m
    enable_notifications = true
    enable_active_checks = true
    enable_passive_checks = true
    enable_event_handler = true
    enable_flapping = true
    enable_perfdata = true
    zone = "client-zone"
    vars.OS = [ "Linux" ]
}

**Comment by Thomas-Gelf** [comment]
Well... the script reflects what your configuration looks like. Did you manually mess with Zone/Endpoint configuration in the Director? The short introduction when going to `Infrastructure` reads:

> When you feel the desire to manually create Zone or Endpoint objects please rethink this twice. Doing so is mostly the wrong way, might lead to a dead end, requiring quite some effort to clean up the whole mess afterwards.

It does so for a reason. If you want Director to assist you with your Agents all you need to do is to toggle the `Use Agent` setting on your host (or Host template). It will then render Zone and Endpoint objects accordingly for you.

**Comment by gms1234** [comment]
I defined the zones and endpoints on my icinga-master. Imported them via the Kickstart Wizard. Defined the ìcinga-client` as agent:
![icinga-client](https://user-images.githubusercontent.com/39398509/41030355-71b4ddfc-697e-11e8-9247-cb0a77cfacff.PNG)

Do you mean I do not have to define them in zones.conf and just add the client under `Hosts`?

**Comment by Thomas-Gelf** [comment]
In case the `icinga-client` endpoint really no longer exists on your master and given that you restarted your master since you removed it: could you please trigger kickstart again? It should remove the Endpoint, and config preview will change. It should show the related Zone/Endpoint as rendered by the Director.

**Comment by gms1234** [comment]
Where did I state that I removed the `icinga-client`? Or are you telling me to delete the endpoint from my zones.conf? I am confused.

Is the method I used not the right one?

**Comment by Thomas-Gelf** [comment]
You're talking about your masters zones.conf, are you? Well... do you want to configure your Agents manually in your text-based Icinga configuration - or with the Director? It's of course also possible to mix those options, but that's usually tricky.

In case this has been done intentionally please share more details on your configuration, what do the related Zone and Endpoint objects in Director and on your master look like? In case you didn't please drop hard-coded agent/zone objects in case you want the Director to do the magic for you. The related [documentation](https://github.com/Icinga/icingaweb2-module-director/blob/master/doc/24-Working-with-agents.md) is a little bit dated, but most of it should still stand true.

I tried to reproduce your issue by manually defining Agent Endpoint and Zone in my Director, but even then - the generated script looks good to me.

**Comment by gms1234** [comment]
Are your master and your client on the same zone? 

When I put the client on the `master-zone` the script looks fine. But this does not resolve my problem.

I followed [this](https://www.youtube.com/watch?v=7eB6pLs9ULE) tutorial. He defines the zones and endpoints in his master's zones.conf and imports it via the Kickstart Wizard.

**Comment by Thomas-Gelf** [comment]
Sorry, I'm unable to understand what you're trying to accomplish. I didn't watch the linked tutorial, but it's title suggests it's talking about adding Satellite Zones for a distributed setup. Until now, in this issue you've been talking about adding an Agent.

**Comment by gms1234** [comment]
You are right. My definition of client/satellite might be wrong. So, change `client` to `satellite`. If I want to add a client - which is on the same zone as my master - the script works fine.

But if I want to add a satellite, which is on a different network, let's say `satellite-zone`, my statement about the script not working correctly stands. It defines the wrong `PARENT_ZONE` and the wrong `PARENT_NODE`.

**Comment by Thomas-Gelf** [comment]
This script is for Agents only. It might partially help with satellites, but it won't help without manual tweaking. I guess #381 is what you're looking for.

**Comment by gms1234** [comment]
I see. Well, this took some time! :D

Again, thank you very much for taking the time!

Should I file a feature request, or is this on your radar now? I really think this is a feature which should be implemented soon.

**Comment by Thomas-Gelf** [comment]
As long as there is an open issue it's on our radar :laughing:

**Comment by lazyfrosch** [comment]
This should be fixed with #1596 - if there is any issue with the new script please open a new issue.
